### PR TITLE
[WIP] Add Polymer 2 support

### PR DESCRIPTION
**Pull request by leonrodenburg**
## What I did

We added initial support for Polymer 2 and set up an example application. We are in the process of improving the compatibility in all browsers and adding support for all Storybook addons. Tests have to be added.

## How to test
* Run `yarn` in the repo root
* Run `yarn bootstrap` in the repo root
* Change to `examples/polymer-cli` directory (`cd examples/polymer-cli`)
* Start Storybook (`yarn storybook`)

The Welcome page and example stories should be shown at `http://localhost:9001/`

## TODO
- [x] Firefox support
- [x] Fix `getstorybook` CLI command for polymer 
- [x] Support for other addons, especially `knobs`
- [x] Fix tests
- [x] Improve examples stories
- [x] Allow string templates
- [x] Add example stories for Polymer 
- [x] HMR
- [x] Try to generalise for all web component frameworks (won't fix in this PR)
- [ ] Cleanup code (lot of React stuff in there from copying)
- [x] Bring PR up to date with latest changes

### CODE REVIEW

**Comment by ndelangen** [comment]
When we have a release of this, we should create a PR here:
https://github.com/Granze/awesome-polymer

**Comment by daKmoR** [comment]
hey I just wanted to git this a quick test but as I never used storybook before I'm a little lost...

what I did
```
git clone https://github.com/storybooks/storybook.git
cd storybook
git checkout add-polymer-support
yarn install
yarn bootstrap
```

what I get
```
? Select which packages to bootstrap
 ◯ Clean and re-install dependencies (reset)
 ◉ Core & Examples (core)
 ◯ Documentation (docs)
 ◯ Build tarballs of packages (build-packs)
 ◯ React-Native example (react-native-vanilla)
❯◯ React-Native-App example (crna-kitchen-sink)
```

so there i no polymer? howto proceed?

**Comment by Hypnosphi** [comment]
It should be included in `core`, I believe

**Comment by leonrodenburg** [comment]
@daKmoR Yes, if you select the `Core & Examples (core)` and have Lerna bootstrap that, the example application in `examples/polymer-cli` should work:

```bash
cd examples/polymer-cli
yarn storybook
```

It is work in progress though, so the feature set is still pretty minimal. The stories in the example app (`src/stories/index.js`) do give a good idea of how Storybook and web components would interact.

**Comment by leonrodenburg** [comment]
@stijnkoopal Nice commits! I'd removed the babel-loader because I was overanalyzing problems I had with a fresh Polymer CLI application + Storybook, but I forgot that it doesn't really matter if your codebase contains Webpack / Babel or not because you import your components in the `src/stories/index.js`, so they are passed through Storybook's WP anyway.

Have you seen the knobs addon? It works pretty cool now!

**Comment by leonrodenburg** [comment]
I still have the problems I was trying to fix earlier, when adding Storybook to a vanilla Polymer CLI app (without Webpack), I get the following error in Chrome:



Removing the custom-elements-es5-adapter.js works in Chrome, but breaks FF. Any ideas?

**Comment by leonrodenburg** [comment]
Update: it seems that in the example app in the repo, the WrapStory.html is properly transpiled into ES5 and therefore works correctly with the custom elements adapter. In a separate repo, it is included in the bundle as-is. Trying to fix now...

**Comment by leonrodenburg** [comment]
💥 Boom! Works now in both Chrome and FF, both the example app with webpack and a vanilla one without, both with and without knobs. I hate Babel, sometimes.

**Comment by stijnkoopal** [comment]
@leonrodenburg Cool! Just added examples for string templates, and html separated from js. Seems to work perfectly. Knobs is really cool, great job

**Comment by ndelangen** [comment]
I will merge in the base-branch today and do a alpha release of this package

**Comment by ndelangen** [comment]
I've released the package `app/polymer` from this branch on npm as version `3.3.0-alpha.3` It's published under the latest tag, because the first release of a package is required to be tagged as latest.

**Comment by ndelangen** [comment]
@Hypnosphi Is there something we can do to fix the cli tests?

**Review by ndelangen — APPROVED** [review]

**Comment by equinusocio** [comment]
Will this work with incoming Polymer 3?

File: examples/polymer-cli/package.json [hunk 1]
```diff
@@ -0,0 +1,31 @@
+{
+  "name": "polymer-cli",
+  "version": "1.0.0",
+  "scripts": {
+    "start": "webpack-dev-server",
+    "storybook": "start-storybook -p 9001 -c .storybook"
+  },
+  "dependencies": {
+    "@polymer/polymer": "^2.2.0",
+    "@storybook/addon-actions": "3.3.0-alpha.3",
```
Comment (Hypnosphi) [inline examples/polymer-cli/package.json]: Please change it to `^3.3.0-alpha.4` here and below

**Review by Hypnosphi — CHANGES_REQUESTED** [review]

**Comment by leonrodenburg** [comment]
@equinusocio Yes, it should work (although it might need minor tweaks). The fact that the Polymer community is (finally) moving to NPM will make our lives a lot easier.

File: examples/polymer-cli/package.json (hunk 1 above)
Comment (leonrodenburg) [reply examples/polymer-cli/package.json]: @Hypnosphi Changing the version numbers of the dependencies causes an `Accessing nonexistent addons channel` error after bootstrapping again. Are there any recent changes in the addons that require other parts of the example code to change?

**Comment by ndelangen** [comment]
We're aiming to support Polymer 3 once it's out of alpha phase. This is something I'd like to discuss with @robdodson or someone else from the polymer core team, on how safe it really is to start supporting Polymer 3 soon.

We like the direction the polymer team has taken deprecating bower and other things!

File: examples/polymer-cli/package.json (hunk 1 above)
Comment (Hypnosphi) [reply examples/polymer-cli/package.json]: The dependencies in `app/polymer should probably be updated as well

Comment (leonrodenburg) [reply examples/polymer-cli/package.json]: Fixed with @ndelangen's merge and version changes

**Comment by robdodson** [comment]
I would wait on supporting Polymer 3 until the team starts announcing a release candidate.

**Comment by equinusocio** [comment]
@robdodson v3 is not just a v2 delivered with npm?

**Review by Hypnosphi — APPROVED** [review]

**Comment by robdodson** [comment]
@equinusocio yes it will be delivered over npm but I think the team is still designing the project and I don't know when they plan to release it. So supporting an early alpha might be difficult because they may make breaking changes along the way.

**Comment by ndelangen** [comment]
We should add a netlify deploy/site for this, so we can demo this!

**Comment by ndelangen** [comment]
@shilman I'd like your approval to merge this

File: app/polymer/src/client/manager/provider.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
+import { location } from 'global';
+import qs from 'qs';
+import React from 'react';
+import { Provider } from '@storybook/ui';
+import addons from '@storybook/addons';
+import createChannel from '@storybook/channel-postmessage';
+import Preview from './preview';
+
+export default class ReactProvider extends Provider {
```
Comment (shilman) [inline app/polymer/src/client/manager/provider.js]: Is this called ReactProvider and not PolymerProvider because it bridging between Polymer and React?

**Review by shilman — APPROVED** [review]
Is there any reason why don't have demo components defined in here? We do in both react/angular (but don't in vue). don't care too heavily either way, but maybe we should be consistent?

File: lib/cli/generators/POLYMER/template/src/stories/index.stories.js [hunk 3]
```diff
@@ -0,0 +1 @@
+// TODO: add demo stories
```
Comment (Hypnosphi) [inline lib/cli/generators/POLYMER/template/src/stories/index.stories.js]: Will we do that?

**Comment by tonai** [comment]
Hi all,

I have tried the following on my existing stencil project (which is derivated from stencil-component-starter):
```shell
npm install --save-dev @storybook/polymer polymer-webpack-loader
```

Then I added the script line in the package.json:
```json
{
  "scripts": {
    "storybook": "start-storybook -p 9001 -c .storybook -s www"
  }
}
```

I created the `.storybook` folder containing a basic `config.js` file:
```js
import { configure } from '@storybook/polymer';
const req = require.context('../src', true, /\.stories\.js$/);
function loadStories() {
  req.keys().forEach((filename) => req(filename))
}

configure(loadStories, module);
```

And a `preview-head.html` with:
```html
<script src="/build/mycomponent.js"></script>
<link rel="stylesheet" type="text/css" href="/build/mycomponent.css">
```

Finally I created a story:
```js
import { storiesOf } from '@storybook/polymer';
const stories = storiesOf('Button', module);
stories.add('using string', () => ('<hec-button>Click here</hec-button>'));
```

But I got the following error:
```
custom-elements-es5-adapter.js:4 Uncaught TypeError: Class constructor  cannot be invoked without 'new'
    at new j (custom-elements-es5-adapter.js:4)
    at renderMain (VM489 preview.bundle.js:27558)
    at renderPreview (VM489 preview.bundle.js:27574)
    at renderUI (VM489 preview.bundle.js:26461)
    at Object.dispatch (VM489 preview.bundle.js:24235)
    at ConfigApi._renderMain (VM489 preview.bundle.js:27342)
    at render (VM489 preview.bundle.js:27293)
    at ConfigApi.configure (VM489 preview.bundle.js:27318)
    at Object.<anonymous> (VM489 preview.bundle.js:26365)
    at Object.defineProperty.value (VM489 preview.bundle.js:26366)
```

This comes from the `custom-elements-es5-adapter.js` script which is included by `@storybook/polymer` (@storybook/polymer/dist/server/iframe.html.ejs).

I have commented out the `custom-elements-es5-adapter.js` and `webcomponents-lite.js` scripts directly in the file and it finally works.

I don't know if I can override the template `iframe.html.ejs` inside my project, because this is clearly a dirty trick to make it works...

**Comment by ndelangen** [comment]
@naipath something you could do here?

**Comment by equinusocio** [comment]
Polymer 3 is on the way, it will be released in weeks...do worth integrating v2?

**Comment by tonai** [comment]
Updating my process with the usage of `@storybook/polymer@4.0.0-alpha4`:

Just changing the first npm install and keep the rest:
```js
npm install --save-dev @storybook/polymer@4.0.0-alpha.4 polymer-webpack-loader lit-html
```

And it works :+1: :smile: 

Just one warning is remaining:
```
WARNING in ./node_modules/@webcomponents/webcomponentsjs/webcomponents-lite.js
Module not found: Error: Can't resolve 'vertx' in './node_modules/@webcomponents/webcomponentsjs'
```

But doesn't seems to be blocking so far.

**Pull request merged**